src/CalculatorUI/Common/AppLifecycleLogger.cs

```diff
@@ -0,0 +1,167 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using System.Diagnostics;
+
+using Windows.ApplicationModel;
+using Windows.ApplicationModel.Background;
+using Windows.ApplicationModel.Core;
+using Windows.Foundation.Diagnostics;
+using Windows.UI.ViewManagement;
+
+namespace CalculatorApp
+{
+    static public partial class Globals
+    {
+#if SEND_DIAGNOSTICS
+        // c.f. WINEVENT_KEYWORD_RESERVED_63-56 0xFF00000000000000 // Bits 63-56 - channel keywords
+        // c.f. WINEVENT_KEYWORD_*              0x00FF000000000000 // Bits 55-48 - system-reserved keywords
+        public const long MICROSOFT_KEYWORD_LEVEL_1 = 0x0000800000000000; // Bit 47
+        public const long MICROSOFT_KEYWORD_LEVEL_2 = 0x0000400000000000;      // Bit 46
+        public const long MICROSOFT_KEYWORD_LEVEL_3 = 0x0000200000000000;     // Bit 45
+        public const long MICROSOFT_KEYWORD_RESERVED_44 = 0x0000100000000000;   // Bit 44 (reserved for future assignment)
+#else
+        // define all Keyword options as 0 when we do not want to upload app diagnostics
+        public const long MICROSOFT_KEYWORD_LEVEL_1 = 0;
+        public const long MICROSOFT_KEYWORD_LEVEL_2 = 0;
+        public const long MICROSOFT_KEYWORD_LEVEL_3 = 0;
+        public const long MICROSOFT_KEYWORD_RESERVED_44 = 0;
+#endif
+    }
+
+    class AppLifecycleLogger
+    {
+        public static AppLifecycleLogger GetInstance()
+        {
+            return s_selfInstance.Value;
+        }
+
+        public bool GetTraceLoggingProviderEnabled()
+        {
+
+            return m_appLifecycleProvider.Enabled;
+
+        }
+
+        public void LaunchUIResponsive()
+        {
+
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            LogAppLifecycleEvent("ModernAppLaunch_UIResponsive", fields);
+
+        }
+
+        public void LaunchVisibleComplete()
+        {
+
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            LogAppLifecycleEvent("ModernAppLaunch_VisibleComplete", fields);
+
+        }
+
+        public void ResumeUIResponsive()
+        {
+
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            LogAppLifecycleEvent("ModernAppResume_UIResponsive", fields);
+
+        }
+
+        public void ResumeVisibleComplete()
+        {
+
+            if (!GetTraceLoggingProviderEnabled())
+                return;
+
+            LoggingFields fields = new LoggingFields();
+            PopulateAppInfo(fields);
+            LogAppLifecycleEvent("ModernAppResume_VisibleComplete", fields);
```

Comment:
LaunchVisibleComplete（), ResumeUIResponsive(), ResumeVisibleComplete(), etc have the same logic except the log event name, encapsulate this logic into a method.

Reply:
Thanks for the comments. These are great points to fix.
However, we'd like schedule such optimization works after migrating C++/CX to C# successfully since we don't want to introduce refactoring risks into the migration works.